In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib
import matplotlib.pyplot as plt
import os
import pandas_datareader as dr
plt.style.use('seaborn-darkgrid')
import plotly.graph_objects as go
import plotly.express as px

In [2]:
os.chdir(r'/Users/marike/Documents/MARIKE/DATA SCIENCE/Final project/Marike FP/Data')

In [3]:
"""indicate in below code line (convert comment to code) the location of Speed Dating Data 
in case of checking the project from another computer"""
# os.chdir("..")

countries_path = r'WB_country_level_data_0.csv'
countries_df = pd.read_csv(countries_path)

## 1.2 COUNTRIES DATASET

In [4]:
countries_df.head()

iso3c region_id country_name income_id           gdp  \
0   ABW       LCN        Aruba       HIC           NaN   
1   AFG       SAS  Afghanistan       LIC  2.141361e+10   
2   AGO       SSF       Angola       LMC  1.030423e+11   
3   ALB       ECS      Albania       UMC  1.347108e+10   
4   AND       ECS      Andorra       HIC  3.319880e+09   

   composition_food_organic_waste_percent  composition_glass_percent  \
0                                     NaN                        NaN   
1                                     NaN                        NaN   
2                                    51.8                        6.7   
3                                    51.4                        4.5   
4                                    31.2                        8.2   

   composition_metal_percent  composition_other_percent  \
0                        NaN                        NaN   
1                        NaN                        NaN   
2                        4.4                      11.50   
3                        4.8                      15.21   
4                        2.6                      11.60   

   composition_paper_cardboard_percent  ...  \
0                                  NaN  ...   
1                                  NaN  ...   
2                                 11.9  ...   
3                                  9.9  ...   
4                                 35.1  ...   

   waste_treatment_controlled_landfill_percent  \
0                                          NaN   
1                                          NaN   
2                                          NaN   
3                                          NaN   
4                                          NaN   

   waste_treatment_incineration_percent  \
0                                   NaN   
1                                   NaN   
2                                   NaN   
3                                   NaN   
4                                  52.1   

   waste_treatment_landfill_unspecified_percent  \
0                                           NaN   
1                                           NaN   
2                                           NaN   
3                                           NaN   
4                                           NaN   

   waste_treatment_open_dump_percent waste_treatment_other_percent  \
0                                NaN                           NaN   
1                                NaN                           NaN   
2                                NaN                           NaN   
3                                NaN                           NaN   
4                                NaN                           NaN   

  waste_treatment_recycling_percent  \
0                              11.0   
1                               NaN   
2                               NaN   
3                               NaN   
4                               NaN   

  waste_treatment_sanitary_landfill_landfill_gas_system_percent  \
0                                                NaN              
1                                                NaN              
2                                                NaN              
3                                                NaN              
4                                                NaN              

  waste_treatment_unaccounted_for_percent  \
0                                    89.0   
1                                     NaN   
2                                     NaN   
3                                     NaN   
4                                    47.9   

  waste_treatment_waterways_marine_percent  where_where_is_this_data_measured  
0                                      NaN                                NaN  
1                                      NaN                              Other  
2                                      NaN                                NaN  
3                                      NaN                Some di

In [5]:
print('Countries rough dataset has the following shape: ',countries_df.shape)

Countries rough dataset has the following shape:  (217, 51)


In [6]:
countries_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 217 entries, 0 to 216
Data columns (total 51 columns):
iso3c                                                                                    217 non-null object
region_id                                                                                217 non-null object
country_name                                                                             216 non-null object
income_id                                                                                217 non-null object
gdp                                                                                      192 non-null float64
composition_food_organic_waste_percent                                                   176 non-null float64
composition_glass_percent                                                                171 non-null float64
composition_metal_percent                                                                170 non-null float64
composition_other_p

Countries dataset has less initial variables than cities rough dataset, only 51 for 217 countries. 41 of these variables are numerical and ten seem to be categorical but again we note that an important number of variables have a very high number of missing values. 

In [7]:
#we check whether dataset has some duplicated row:
print('Countries dataset has ', countries_df.duplicated().sum(),' duplicated rows.')

Countries dataset has  0  duplicated rows.


In [8]:
#we check missing information quantity.
from MC_FP_UsefulFunctions import *

countries_null_perc = show_null(countries_df)
countries_null_perc

Null data amount  \
waste_collection_coverage_rural_percent_of_geog...               216   
waste_collection_coverage_total_percent_of_geog...               216   
waste_collection_coverage_urban_percent_of_geog...               216   
waste_treatment_anaerobic_digestion_percent                      215   
waste_treatment_waterways_marine_percent                         213   
waste_collection_coverage_rural_percent_of_waste                 209   
waste_collection_coverage_urban_percent_of_waste                 208   
waste_collection_coverage_rural_percent_of_hous...               208   
waste_collection_coverage_urban_percent_of_hous...               208   
waste_collection_coverage_total_percent_of_hous...               205   
waste_collection_coverage_rural_percent_of_popu...               194   
waste_treatment_other_percent                                    192   
waste_treatment_sanitary_landfill_landfill_gas_...               189   
composition_yard_garden_green_waste_percent                      179   
waste_collection_coverage_urban_percent_of_popu...               177   
special_waste_industrial_waste_tons_year                         177   
waste_treatment_controlled_landfill_percent                      174   
waste_treatment_incineration_percent                             169   
waste_collection_coverage_total_percent_of_waste                 167   
composition_rubber_leather_percent                               164   
special_waste_agricultural_waste_tons_year                       164   
where_where_is_this_data_measured                                160   
waste_treatment_landfill_unspecified_percent                     154   
composition_wood_percent                                         153   
waste_treatment_open_dump_percent                                149   
waste_treatment_compost_percent                                  149   
other_information_summary_of_key_solid_waste_in...               144   
special_waste_construction_and_demolition_waste...               143   
waste_collection_coverage_total_percent_of_popu...               125   
waste_treatment_unaccounted_for_percent                          123   
special_waste_medical_waste_tons_year                            112   
other_information_information_system_for_solid_...               105   
special_waste_hazardous_waste_tons_year                           99   
waste_treatment_recycling_percent                                 93   
other_information_ppp_rules_and_regulations                       55   
other_information_national_agency_to_enforce_so...                53   
composition_metal_percent                                         47   
composition_glass_percent                                         46   
composition_plastic_percent                                       42   
composition_other_percent                                         42   
composition_food_organic_waste_percent                            41   
composition_paper_cardboard_percent                               41   
special_waste_e_waste_tons_year                                   35   
gdp                                                               25   
other_information_national_law_governing_solid_...                23   
total_msw_total_msw_generated_tons_year                            2   
country_name                                                       1   
income_id                                                          0   
population_population_number_of_people                             0   
region_id                                                          0   
iso3c                                                              0   

                                                    Missing data (%)     type  
waste_collection_coverage_rural_percent_of_geog...         99.539171  float64  
waste_collection_coverage_total_percent_of_geog...         99.539171  float64  
waste_collection_coverage_urban_percent_of_geog...         99.53917

In [9]:
print('Number of variables with more than 90% missing information: ', len(countries_null_perc.loc[countries_null_perc['Missing data (%)']>90]), '. That is to say ',round(len(countries_null_perc.loc[countries_null_perc['Missing data (%)']>90])*100/len(countries_null_perc),2),'% of the variables')
print('Number of variables with more than 75% missing information: ', len(countries_null_perc.loc[countries_null_perc['Missing data (%)']>75]), '. That is to say ',round(len(countries_null_perc.loc[countries_null_perc['Missing data (%)']>75])*100/len(countries_null_perc),2),'% of the variables')
print('Number of variables with more than 50% missing information: ', len(countries_null_perc.loc[countries_null_perc['Missing data (%)']>50]), '. That is to say ',round(len(countries_null_perc.loc[countries_null_perc['Missing data (%)']>50])*100/len(countries_null_perc),2),'% of the variables')
print('Number of variables with more than 40% missing information: ', len(countries_null_perc.loc[countries_null_perc['Missing data (%)']>40]), '. That is to say ',round(len(countries_null_perc.loc[countries_null_perc['Missing data (%)']>40])*100/len(countries_null_perc),2),'% of the variables')

Number of variables with more than 90% missing information:  10 . That is to say  19.61 % of the variables
Number of variables with more than 75% missing information:  21 . That is to say  41.18 % of the variables
Number of variables with more than 50% missing information:  31 . That is to say  60.78 % of the variables
Number of variables with more than 40% missing information:  34 . That is to say  66.67 % of the variables


In [10]:
countries_df.region_id.unique()

array(['LCN', 'SAS', 'SSF', 'ECS', 'MEA', 'EAS', 'NAC'], dtype=object)

In [11]:
countries_df.income_id.unique()

array(['HIC', 'LIC', 'LMC', 'UMC'], dtype=object)

In [12]:
countries_df.replace({'LIC': 'LowIncome', 'LMC': 'LowMiddleIncome','UMC':'MiddleUpperIncome','HIC':'HighIncome'},inplace=True)

In [13]:
countries_df.income_id.unique()

array(['HighIncome', 'LowIncome', 'LowMiddleIncome', 'MiddleUpperIncome'],
      dtype=object)

Countries have the variables *region_id* and *income_id* in commmon with cities. So again, before dropping massively information, we will check if the missing information is so important for each sub geographical or economic area.

In [14]:
Countries_null_perc_X_region_income = pd.DataFrame()

Countries_null_perc_X_region_income['World']=countries_null_perc['Missing data (%)']

for x in countries_df.region_id.unique():
    null_perc_country_region = show_null(countries_df.loc[countries_df.region_id == x])
    Countries_null_perc_X_region_income[x]= round(null_perc_country_region['Missing data (%)'],2)

for x in countries_df.income_id.unique():
    null_perc_country_income = show_null(countries_df.loc[countries_df.income_id == x])
    Countries_null_perc_X_region_income[x]= round(null_perc_country_income['Missing data (%)'],2)    

Countries_null_perc_X_region_income

World     LCN    SAS  \
waste_collection_coverage_rural_percent_of_geog...  99.539171  100.00  100.0   
waste_collection_coverage_total_percent_of_geog...  99.539171  100.00  100.0   
waste_collection_coverage_urban_percent_of_geog...  99.539171  100.00  100.0   
waste_treatment_anaerobic_digestion_percent         99.078341  100.00  100.0   
waste_treatment_waterways_marine_percent            98.156682   97.62   87.5   
waste_collection_coverage_rural_percent_of_waste    96.313364  100.00   75.0   
waste_collection_coverage_urban_percent_of_waste    95.852535  100.00   50.0   
waste_collection_coverage_rural_percent_of_hous...  95.852535   85.71  100.0   
waste_collection_coverage_urban_percent_of_hous...  95.852535   88.10  100.0   
waste_collection_coverage_total_percent_of_hous...  94.470046   80.95  100.0   
waste_collection_coverage_rural_percent_of_popu...  89.400922   92.86  100.0   
waste_treatment_other_percent                       88.479263   80.95   87.5   
waste_treatment_sanitary_landfill_landfill_gas_...  87.096774   61.90  100.0   
composition_yard_garden_green_waste_percent         82.488479   85.71   62.5   
waste_collection_coverage_urban_percent_of_popu...  81.566820   85.71   87.5   
special_waste_industrial_waste_tons_year            81.566820  100.00   62.5   
waste_treatment_controlled_landfill_percent         80.184332   50.00   87.5   
waste_treatment_incineration_percent                77.880184   95.24   87.5   
waste_collection_coverage_total_percent_of_waste    76.958525   90.48   75.0   
composition_rubber_leather_percent                  75.576037   83.33   62.5   
special_waste_agricultural_waste_tons_year          75.576037  100.00   62.5   
where_where_is_this_data_measured                   73.732719   64.29   12.5   
waste_treatment_landfill_unspecified_percent        70.967742   90.48   75.0   
composition_wood_percent                            70.506912   80.95   75.0   
waste_treatment_open_dump_percent                   68.663594   50.00   62.5   
waste_treatment_compost_percent                     68.663594   80.95   12.5   
other_information_summary_of_key_solid_waste_in...  66.359447   52.38   50.0   
special_waste_construction_and_demolition_waste...  65.898618   90.48   62.5   
waste_collection_coverage_total_percent_of_popu...  57.603687   38.10   87.5   
waste_treatment_unaccounted_for_percent             56.682028   45.24   75.0   
special_waste_medical_waste_tons_year               51.612903   69.05    0.0   
other_information_information_system_for_solid_...  48.387097   35.71   25.0   
special_waste_hazardous_waste_tons_year             45.622120   52.38   37.5   
waste_treatment_recycling_percent                   42.857143   57.14   50.0   
other_information_ppp_rules_and_regulations         25.345622   19.05    0.0   
other_information_national_agency_to_enforce_so...  24.423963   26.19    0.0   
composition_metal_percent                           21.658986   16.67   25.0   
composition_glass_percent                           21.198157   14.29   37.5   
composition_plastic_percent                         19.354839   14.29   25.0   
composition_other_percent                           19.354839   14.29   25.0   
composition_food_organic_waste_percent              18.894009   14.29   25.0   
composition_paper_cardboard_percent                 18.894009   16.67   25.0   
special_waste_e_waste_tons_year                     16.129032   23.81    0.0   
gdp                                                 11.520737   23.81    0.0   
other_information_national_law_governing_solid_...  10.599078    0.00    0.0   
total_msw_total_msw_generated_tons_year              0.921659    4.76    0.0   
country_name                                         0.460829    0.00    0.0   
income_id                                            0.000000    0.00    0.0   
population_population_number_of_people               0.000000    0.00    0.0   
region_id                                            0.0

AS with cities dataset, we can see that some variables have a very high percentage of missing information for all regions/income type. We keep same criteria as previously with cities: we decide to keep only variables with less than 50% of missing information except for variables with 50 to 70% missing information but for which there is a world area (defined geographically or by income) where this variable has less than 30% missing information.


In [15]:
Countries_Notallregions_Income_HighNaNPerc=[]

for x in Countries_null_perc_X_region_income.index:
    
    if Countries_null_perc_X_region_income.at[x,'World']<70 and Countries_null_perc_X_region_income.at[x,'World']>50:
        
        for y in countries_df.region_id.unique():
            if Countries_null_perc_X_region_income.at[x,y]<30:
                Countries_Notallregions_Income_HighNaNPerc.append(x)
                
        for z in countries_df.income_id.unique():
            if Countries_null_perc_X_region_income.at[x,z]<30:
                Countries_Notallregions_Income_HighNaNPerc.append(x)
                
Countries_Notallregions_Income_HighNaNPerc


['waste_treatment_compost_percent',
 'special_waste_construction_and_demolition_waste_tons_year',
 'waste_collection_coverage_total_percent_of_population',
 'special_waste_medical_waste_tons_year',
 'special_waste_medical_waste_tons_year',
 'special_waste_medical_waste_tons_year']

In [16]:
Countries_Notallregions_Income_HighNaNPerc_df = pd.DataFrame({'Variable to keep':Countries_Notallregions_Income_HighNaNPerc})

In [17]:
Countries_Notallregions_Income_HighNaNPerc_df['Variable to keep'].unique()

array(['waste_treatment_compost_percent',
       'special_waste_construction_and_demolition_waste_tons_year',
       'waste_collection_coverage_total_percent_of_population',
       'special_waste_medical_waste_tons_year'], dtype=object)

In [18]:
Countries_ColToDrop = []
for x in Countries_null_perc_X_region_income.index:
    if Countries_null_perc_X_region_income.at[x,'World']>50:
        if x not in Countries_Notallregions_Income_HighNaNPerc_df['Variable to keep'].unique():
            Countries_ColToDrop.append(x)

Countries_ColToDrop

['waste_collection_coverage_rural_percent_of_geographic_area',
 'waste_collection_coverage_total_percent_of_geographic_area',
 'waste_collection_coverage_urban_percent_of_geographic_area',
 'waste_treatment_anaerobic_digestion_percent',
 'waste_treatment_waterways_marine_percent',
 'waste_collection_coverage_rural_percent_of_waste',
 'waste_collection_coverage_urban_percent_of_waste',
 'waste_collection_coverage_rural_percent_of_households',
 'waste_collection_coverage_urban_percent_of_households',
 'waste_collection_coverage_total_percent_of_households',
 'waste_collection_coverage_rural_percent_of_population',
 'waste_treatment_other_percent',
 'waste_treatment_sanitary_landfill_landfill_gas_system_percent',
 'composition_yard_garden_green_waste_percent',
 'waste_collection_coverage_urban_percent_of_population',
 'special_waste_industrial_waste_tons_year',
 'waste_treatment_controlled_landfill_percent',
 'waste_treatment_incineration_percent',
 'waste_collection_coverage_total_percen

In [19]:
countries_df_1 = countries_df.drop(Countries_ColToDrop, axis=1)
countries_df_1

iso3c region_id  country_name          income_id           gdp  \
0     ABW       LCN         Aruba         HighIncome           NaN   
1     AFG       SAS   Afghanistan          LowIncome  2.141361e+10   
2     AGO       SSF        Angola    LowMiddleIncome  1.030423e+11   
3     ALB       ECS       Albania  MiddleUpperIncome  1.347108e+10   
4     AND       ECS       Andorra         HighIncome  3.319880e+09   
..    ...       ...           ...                ...           ...   
212   XKX       ECS        Kosovo    LowMiddleIncome  7.129272e+09   
213   YEM       MEA   Yemen, Rep.          LowIncome  1.192703e+10   
214   ZAF       SSF  South Africa  MiddleUpperIncome  4.212087e+11   
215   ZMB       SSF        Zambia    LowMiddleIncome  2.703717e+10   
216   ZWE       SSF      Zimbabwe          LowIncome  1.481899e+10   

     composition_food_organic_waste_percent  composition_glass_percent  \
0                                       NaN                        NaN   
1                                       NaN                        NaN   
2                                 51.800000                   6.700000   
3                                 51.400000                   4.500000   
4                                 31.200000                   8.200000   
..                                      ...                        ...   
212                               42.000000                   6.000000   
213                               65.000000                   1.000000   
214                               16.381655                   5.200216   
215                                     NaN                        NaN   
216                               36.000000                   5.000000   

     composition_metal_percent  composition_other_percent  \
0                          NaN                        NaN   
1                          NaN                        NaN   
2                     4.400000                  11.500000   
3                     4.800000                  15.210000   
4                     2.600000                  11.600000   
..                         ...                        ...   
212                   6.000000                  20.000000   
213                   6.000000                   6.000000   
214                  16.910461                  45.020646   
215                        NaN                        NaN   
216                   6.000000                   3.000000   

     composition_paper_cardboard_percent  ...  \
0                                    NaN  ...   
1                                    NaN  ...   
2                              11.900000  ...   
3                               9.900000  ...   
4                              35.100000  ...   
..                                   ...  ...   
212                             8.000000  ...   
213                             7.000000  ...   
214                             9.396918  ...   
215                                  NaN  ...   
216                            27.000000  ...   

     other_information_ppp_rules_and_regulations  \
0                                            Yes   
1                                            Yes   
2                                            Yes   
3                                            Yes   
4                                            Yes   
..                                           ...   
212                                          Yes   
213                                           No   
214                                          NaN   
215                                          NaN   
216                                          NaN   

    population_population_number_of_people  \
0                                103187.00   
1                              34656032.00   
2                              25096150.00   
3                               2880703.00   
4                                 82431.00   
..                                     ...   
212  

In [20]:
countries_df_1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 217 entries, 0 to 216
Data columns (total 24 columns):
iso3c                                                                             217 non-null object
region_id                                                                         217 non-null object
country_name                                                                      216 non-null object
income_id                                                                         217 non-null object
gdp                                                                               192 non-null float64
composition_food_organic_waste_percent                                            176 non-null float64
composition_glass_percent                                                         171 non-null float64
composition_metal_percent                                                         170 non-null float64
composition_other_percent                                                  

In [21]:
#We check missing information for catgorical type varaibles.
for x in countries_df_1.columns:
    if countries_df_1[x].dtypes==object:
        if x not in ['iso3c','region_id','country_name','city_name']:
            print (x,' , ', countries_df_1[x].unique())

income_id  ,  ['HighIncome' 'LowIncome' 'LowMiddleIncome' 'MiddleUpperIncome']
other_information_information_system_for_solid_waste_management  ,  [nan 'No' 'Yes' 'None']
other_information_national_agency_to_enforce_solid_waste_laws_and_regulations  ,  ['Yes' nan 'No' 'None']
other_information_national_law_governing_solid_waste_management_in_the_country  ,  ['Yes' nan 'No' 'yes']
other_information_ppp_rules_and_regulations  ,  ['Yes' 'No' nan 'None']


Again we notice that among the categorical variables, some have binary variables with NO or YES values but in some cases null information is considered instead of NO or the variables have unappropriate repeated values because of lower or upper case writing. We correct this:¶

In [22]:
countries_df_1.replace({'Yes':'yes','No':'no','None':'no'},inplace=True)

In [23]:
Countries_VariableReplaceNanbyNo = []
for x in countries_df_1.columns:
    if countries_df_1[x].dtypes==object:
        if 'yes' in countries_df_1[x].unique():
            Countries_VariableReplaceNanbyNo.append(x)

Countries_VariableReplaceNanbyNo   

['other_information_information_system_for_solid_waste_management',
 'other_information_national_agency_to_enforce_solid_waste_laws_and_regulations',
 'other_information_national_law_governing_solid_waste_management_in_the_country',
 'other_information_ppp_rules_and_regulations']

In [24]:
for x in Countries_VariableReplaceNanbyNo:
    countries_df_1[x].fillna('no', inplace=True)

countries_df_1.isna().sum()

iso3c                                                                               0
region_id                                                                           0
country_name                                                                        1
income_id                                                                           0
gdp                                                                                25
composition_food_organic_waste_percent                                             41
composition_glass_percent                                                          46
composition_metal_percent                                                          47
composition_other_percent                                                          42
composition_paper_cardboard_percent                                                41
composition_plastic_percent                                                        42
other_information_information_system_for_solid_waste_m

In [25]:
countries_df_1.columns

Index(['iso3c', 'region_id', 'country_name', 'income_id', 'gdp',
       'composition_food_organic_waste_percent', 'composition_glass_percent',
       'composition_metal_percent', 'composition_other_percent',
       'composition_paper_cardboard_percent', 'composition_plastic_percent',
       'other_information_information_system_for_solid_waste_management',
       'other_information_national_agency_to_enforce_solid_waste_laws_and_regulations',
       'other_information_national_law_governing_solid_waste_management_in_the_country',
       'other_information_ppp_rules_and_regulations',
       'population_population_number_of_people',
       'special_waste_construction_and_demolition_waste_tons_year',
       'special_waste_e_waste_tons_year',
       'special_waste_hazardous_waste_tons_year',
       'special_waste_medical_waste_tons_year',
       'total_msw_total_msw_generated_tons_year',
       'waste_collection_coverage_total_percent_of_population',
       'waste_treatment_compost_percent

In [26]:
countries_df_1 = countries_df_1.rename(columns={'composition_food_organic_waste_percent':'compo_organic',
'composition_glass_percent':'compo_glass',
'composition_metal_percent':'compo_metal',
'composition_other_percent':'compo_other',
'composition_paper_cardboard_percent':'compo_paper_cardboard',
'composition_plastic_percent':'compo_plastic',
'other_information_information_system_for_solid_waste_management':'info_system_SW_mgmt',
'other_information_national_agency_to_enforce_solid_waste_laws_and_regulations':'national_waste_agency',
'other_information_national_law_governing_solid_waste_management_in_the_country':'national_law_for_SW_mgmt',
'other_information_ppp_rules_and_regulations':'PPP_rules_and_regulations',
'population_population_number_of_people':'population',
'special_waste_construction_and_demolition_waste_tons_year':'SpW_construction_waste_tons_year',
'special_waste_e_waste_tons_year':'SpW_e_waste_tons_year',
'special_waste_hazardous_waste_tons_year':'SpW_hazardous_waste_tons_year',
'special_waste_medical_waste_tons_year':'SpW_medical_waste_tons_year',
'total_msw_total_msw_generated_tons_year':'MSW_generated_tons_year',
'waste_collection_coverage_total_percent_of_population':'waste_collection_coverage_perc_pop',
'waste_treatment_compost_percent':'compost_percent',
'waste_treatment_recycling_percent':'recycling_percent'
})
countries_df_1.head()

iso3c region_id country_name          income_id           gdp  \
0   ABW       LCN        Aruba         HighIncome           NaN   
1   AFG       SAS  Afghanistan          LowIncome  2.141361e+10   
2   AGO       SSF       Angola    LowMiddleIncome  1.030423e+11   
3   ALB       ECS      Albania  MiddleUpperIncome  1.347108e+10   
4   AND       ECS      Andorra         HighIncome  3.319880e+09   

   compo_organic  compo_glass  compo_metal  compo_other  \
0            NaN          NaN          NaN          NaN   
1            NaN          NaN          NaN          NaN   
2           51.8          6.7          4.4        11.50   
3           51.4          4.5          4.8        15.21   
4           31.2          8.2          2.6        11.60   

   compo_paper_cardboard  ...  PPP_rules_and_regulations  population  \
0                    NaN  ...                        yes    103187.0   
1                    NaN  ...                        yes  34656032.0   
2                   11.9  ...                        yes  25096150.0   
3                    9.9  ...                        yes   2880703.0   
4                   35.1  ...                        yes     82431.0   

  SpW_construction_waste_tons_year SpW_e_waste_tons_year  \
0                              NaN                   NaN   
1                              NaN               20000.0   
2                              NaN               92000.0   
3                         203009.0               20000.0   
4                              NaN                   NaN   

  SpW_hazardous_waste_tons_year  SpW_medical_waste_tons_year  \
0                           NaN                          NaN   
1                           NaN                      9855.00   
2                           NaN                      5495.95   
3                       5714.82                       114.30   
4                       1788.00                          NaN   

   MSW_generated_tons_year  waste_collection_coverage_perc_pop  \
0             8.813202e+04                                 NaN   
1             5.628525e+06                                 NaN   
2             4.213644e+06                                23.1   
3             1.142964e+06                                60.0   
4             4.300000e+04                               100.0   

   compost_percent  recycling_percent  
0              NaN               11.0  
1              NaN                NaN  
2              NaN                NaN  
3              NaN                NaN  
4              NaN                NaN  

[5 rows x 24 columns]

In [27]:
countries_df_1.loc[countries_df_1['MSW_generated_tons_year'].isna()] 

iso3c region_id               country_name   income_id  gdp  \
183   SXM       LCN  Sint Maarten (Dutch part)  HighIncome  NaN   
186   TCA       LCN   Turks and Caicos Islands  HighIncome  NaN   

     compo_organic  compo_glass  compo_metal  compo_other  \
183           46.0          7.0          7.0         12.0   
186           21.8         20.5          7.6          2.4   

     compo_paper_cardboard  ...  PPP_rules_and_regulations population  \
183                   15.0  ...                         no    37685.0   
186                   31.5  ...                        yes    34900.0   

    SpW_construction_waste_tons_year SpW_e_waste_tons_year  \
183                              NaN                   NaN   
186                              NaN                   NaN   

    SpW_hazardous_waste_tons_year  SpW_medical_waste_tons_year  \
183                           NaN                          NaN   
186                           NaN                          NaN   

     MSW_generated_tons_year  waste_collection_coverage_perc_pop  \
183                      NaN                                 NaN   
186                      NaN                                 NaN   

     compost_percent  recycling_percent  
183              NaN                NaN  
186              NaN                NaN  

[2 rows x 24 columns]

In [28]:
# Since the 2 rows also have missing GDP information and are for small islands we discard them.
countries_df_2 = countries_df_1.dropna(subset = ['MSW_generated_tons_year']) 

In [29]:
countries_df_3 = countries_df_2.drop(['country_name'],axis=1)
countries_df_3 = countries_df_3.rename(columns={'iso3c': 'country_code'})
countries_df_3.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 215 entries, 0 to 216
Data columns (total 23 columns):
country_code                          215 non-null object
region_id                             215 non-null object
income_id                             215 non-null object
gdp                                   192 non-null float64
compo_organic                         174 non-null float64
compo_glass                           169 non-null float64
compo_metal                           168 non-null float64
compo_other                           173 non-null float64
compo_paper_cardboard                 174 non-null float64
compo_plastic                         173 non-null float64
info_system_SW_mgmt                   215 non-null object
national_waste_agency                 215 non-null object
national_law_for_SW_mgmt              215 non-null object
PPP_rules_and_regulations             215 non-null object
population                            215 non-null float64
SpW_construction_

In [30]:
countries_df_3.isna().sum()

country_code                            0
region_id                               0
income_id                               0
gdp                                    23
compo_organic                          41
compo_glass                            46
compo_metal                            47
compo_other                            42
compo_paper_cardboard                  41
compo_plastic                          42
info_system_SW_mgmt                     0
national_waste_agency                   0
national_law_for_SW_mgmt                0
PPP_rules_and_regulations               0
population                              0
SpW_construction_waste_tons_year      141
SpW_e_waste_tons_year                  33
SpW_hazardous_waste_tons_year          97
SpW_medical_waste_tons_year           110
MSW_generated_tons_year                 0
waste_collection_coverage_perc_pop    123
compost_percent                       147
recycling_percent                      91
dtype: int64

In order not to loose the information in the rows corresponding to missing gdp, since income_id and gdp are related, when gdp is missing we want to replace it by the mean of gdp for the countries with same income_id.
Nevertheless GDP value is also related to population so in order to realize this operation we first create a new variable : GDP by inhabitant (gdp_inhab) which will not depend on country size.

In [31]:
countries_df_3['gdp_inhab']=round(countries_df_3['gdp']/countries_df_3['population'],2)

In [32]:
mean_gdp_inhab_by_income_dict = {}
for x in countries_df_3.income_id.unique():
    mean_gdp_inhab_by_income_dict.update({x:countries_df_3.gdp_inhab.loc[countries_df_3.income_id == x].mean()}) 

mean_gdp_inhab_by_income_dict

{'HighIncome': 35895.51301587302,
 'LowIncome': 743.6455172413795,
 'LowMiddleIncome': 2566.5491304347825,
 'MiddleUpperIncome': 7332.349814814818}

In [33]:
mean_gdp_inhab_by_income_dict['MiddleUpperIncome']

7332.349814814818

In [34]:
for x in countries_df_3.income_id.unique():
 print('number of countries with gdp not informed for countries with ', x, ': ' ,countries_df_3.gdp_inhab.loc[countries_df_3.income_id == x].isna().sum())

number of countries with gdp not informed for countries with  HighIncome :  16
number of countries with gdp not informed for countries with  LowIncome :  4
number of countries with gdp not informed for countries with  LowMiddleIncome :  1
number of countries with gdp not informed for countries with  MiddleUpperIncome :  2


In [35]:
for x in countries_df_3.income_id.unique():
    countries_df_3['gdp_inhab'].loc[countries_df_3.income_id == x]=countries_df_3['gdp_inhab'].loc[countries_df_3.income_id == x].fillna(mean_gdp_inhab_by_income_dict[x])

/Users/marike/opt/anaconda3/lib/python3.7/site-packages/pandas/core/indexing.py:205: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



In [36]:
countries_df_3['gdp_inhab'].isna().sum()

0

In [37]:
countries_df_3[['country_code','gdp_inhab']].loc[countries_df_3['country_code']=='CUB']

country_code    gdp_inhab
45          CUB  7332.349815

In [38]:
countries_df_3[['country_code','gdp_inhab']]

country_code     gdp_inhab
0            ABW  35895.513016
1            AFG    617.890000
2            AGO   4105.900000
3            ALB   4676.320000
4            AND  40274.660000
..           ...           ...
212          XKX   3956.750000
213          YEM    432.390000
214          ZAF   8142.550000
215          ZMB   1895.380000
216          ZWE   1185.470000

[215 rows x 2 columns]

As with Cities, we drop the variable *waste_collection_coverage_total_percent_of_population* because we consider it is not useful for our study.

In [39]:
countries_df_4 = countries_df_3.drop(['gdp','waste_collection_coverage_perc_pop'],axis=1)

In [40]:
countries_df_4[['compost_percent','recycling_percent']] = countries_df_4[['compost_percent','recycling_percent']].fillna(0)

In [41]:
countries_df_4.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 215 entries, 0 to 216
Data columns (total 22 columns):
country_code                        215 non-null object
region_id                           215 non-null object
income_id                           215 non-null object
compo_organic                       174 non-null float64
compo_glass                         169 non-null float64
compo_metal                         168 non-null float64
compo_other                         173 non-null float64
compo_paper_cardboard               174 non-null float64
compo_plastic                       173 non-null float64
info_system_SW_mgmt                 215 non-null object
national_waste_agency               215 non-null object
national_law_for_SW_mgmt            215 non-null object
PPP_rules_and_regulations           215 non-null object
population                          215 non-null float64
SpW_construction_waste_tons_year    74 non-null float64
SpW_e_waste_tons_year               182 non-null

In [42]:
waste_compo_var = countries_df_4.filter(regex='compo_').columns.tolist()
waste_compo_var

['compo_organic',
 'compo_glass',
 'compo_metal',
 'compo_other',
 'compo_paper_cardboard',
 'compo_plastic']

In [43]:
special_waste_var = countries_df_4.filter(regex= 'SpW').columns.tolist()
special_waste_var

['SpW_construction_waste_tons_year',
 'SpW_e_waste_tons_year',
 'SpW_hazardous_waste_tons_year',
 'SpW_medical_waste_tons_year']

As we have done previously for cities, in order to be able to deal more ahead with the information about recycling percentage, compost percentage. waste composition and special waste types, we extract this information when available in sub datasets for further use and then we remove it from the main dataset.


In [44]:
countries_construction_waste = countries_df_4.loc[countries_df_4.SpW_construction_waste_tons_year.notnull()]
countries_construction_waste.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 74 entries, 3 to 212
Data columns (total 22 columns):
country_code                        74 non-null object
region_id                           74 non-null object
income_id                           74 non-null object
compo_organic                       69 non-null float64
compo_glass                         66 non-null float64
compo_metal                         66 non-null float64
compo_other                         67 non-null float64
compo_paper_cardboard               70 non-null float64
compo_plastic                       69 non-null float64
info_system_SW_mgmt                 74 non-null object
national_waste_agency               74 non-null object
national_law_for_SW_mgmt            74 non-null object
PPP_rules_and_regulations           74 non-null object
population                          74 non-null float64
SpW_construction_waste_tons_year    74 non-null float64
SpW_e_waste_tons_year               73 non-null float64
SpW_haz

In [45]:
SpW_except_construction_var = ['SpW_e_waste_tons_year',
 'SpW_hazardous_waste_tons_year',
 'SpW_medical_waste_tons_year']
countries_construction_waste_df = countries_construction_waste.drop((waste_compo_var+SpW_except_construction_var+['compost_percent','recycling_percent']),axis=1)

In [46]:
countries_construction_waste_df .info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 74 entries, 3 to 212
Data columns (total 11 columns):
country_code                        74 non-null object
region_id                           74 non-null object
income_id                           74 non-null object
info_system_SW_mgmt                 74 non-null object
national_waste_agency               74 non-null object
national_law_for_SW_mgmt            74 non-null object
PPP_rules_and_regulations           74 non-null object
population                          74 non-null float64
SpW_construction_waste_tons_year    74 non-null float64
MSW_generated_tons_year             74 non-null float64
gdp_inhab                           74 non-null float64
dtypes: float64(4), object(7)
memory usage: 6.9+ KB


In [47]:
countries_e_waste = countries_df_4.loc[countries_df_4['SpW_e_waste_tons_year'].notnull()]

In [48]:
SpW_except_e_waste = ['SpW_construction_waste_tons_year',
 'SpW_hazardous_waste_tons_year','SpW_medical_waste_tons_year']

In [49]:
countries_e_waste_df = countries_e_waste.drop((waste_compo_var+SpW_except_e_waste+['compost_percent','recycling_percent']),axis=1)
countries_e_waste_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 182 entries, 1 to 216
Data columns (total 11 columns):
country_code                 182 non-null object
region_id                    182 non-null object
income_id                    182 non-null object
info_system_SW_mgmt          182 non-null object
national_waste_agency        182 non-null object
national_law_for_SW_mgmt     182 non-null object
PPP_rules_and_regulations    182 non-null object
population                   182 non-null float64
SpW_e_waste_tons_year        182 non-null float64
MSW_generated_tons_year      182 non-null float64
gdp_inhab                    182 non-null float64
dtypes: float64(4), object(7)
memory usage: 17.1+ KB


In [50]:
countries_hazardous_waste = countries_df_4.loc[countries_df_4['SpW_hazardous_waste_tons_year'].notnull()]
SpW_except_haz_waste = ['SpW_construction_waste_tons_year',
 'SpW_e_waste_tons_year', 'SpW_medical_waste_tons_year']
countries_hazardous_waste_df = countries_hazardous_waste.drop((waste_compo_var+SpW_except_haz_waste+['compost_percent','recycling_percent']),axis=1)
countries_hazardous_waste_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 118 entries, 3 to 216
Data columns (total 11 columns):
country_code                     118 non-null object
region_id                        118 non-null object
income_id                        118 non-null object
info_system_SW_mgmt              118 non-null object
national_waste_agency            118 non-null object
national_law_for_SW_mgmt         118 non-null object
PPP_rules_and_regulations        118 non-null object
population                       118 non-null float64
SpW_hazardous_waste_tons_year    118 non-null float64
MSW_generated_tons_year          118 non-null float64
gdp_inhab                        118 non-null float64
dtypes: float64(4), object(7)
memory usage: 11.1+ KB


In [51]:
countries_medical_waste = countries_df_4.loc[countries_df_4['SpW_medical_waste_tons_year'].notnull()]
SpW_except_med_waste = ['SpW_construction_waste_tons_year',
 'SpW_e_waste_tons_year','SpW_hazardous_waste_tons_year']
countries_medical_waste_df = countries_medical_waste.drop((waste_compo_var+SpW_except_med_waste+['compost_percent','recycling_percent']),axis=1)
countries_medical_waste_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 105 entries, 1 to 212
Data columns (total 11 columns):
country_code                   105 non-null object
region_id                      105 non-null object
income_id                      105 non-null object
info_system_SW_mgmt            105 non-null object
national_waste_agency          105 non-null object
national_law_for_SW_mgmt       105 non-null object
PPP_rules_and_regulations      105 non-null object
population                     105 non-null float64
SpW_medical_waste_tons_year    105 non-null float64
MSW_generated_tons_year        105 non-null float64
gdp_inhab                      105 non-null float64
dtypes: float64(4), object(7)
memory usage: 9.8+ KB


In [52]:
countries_waste_composition = countries_df_4.dropna(subset = waste_compo_var) 
print(countries_waste_composition.shape)
countries_waste_composition.info()

(159, 22)
<class 'pandas.core.frame.DataFrame'>
Int64Index: 159 entries, 2 to 216
Data columns (total 22 columns):
country_code                        159 non-null object
region_id                           159 non-null object
income_id                           159 non-null object
compo_organic                       159 non-null float64
compo_glass                         159 non-null float64
compo_metal                         159 non-null float64
compo_other                         159 non-null float64
compo_paper_cardboard               159 non-null float64
compo_plastic                       159 non-null float64
info_system_SW_mgmt                 159 non-null object
national_waste_agency               159 non-null object
national_law_for_SW_mgmt            159 non-null object
PPP_rules_and_regulations           159 non-null object
population                          159 non-null float64
SpW_construction_waste_tons_year    63 non-null float64
SpW_e_waste_tons_year               14

In [53]:
countries_waste_composition_df = countries_waste_composition.drop(special_waste_var, axis=1)
countries_waste_composition_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 159 entries, 2 to 216
Data columns (total 18 columns):
country_code                 159 non-null object
region_id                    159 non-null object
income_id                    159 non-null object
compo_organic                159 non-null float64
compo_glass                  159 non-null float64
compo_metal                  159 non-null float64
compo_other                  159 non-null float64
compo_paper_cardboard        159 non-null float64
compo_plastic                159 non-null float64
info_system_SW_mgmt          159 non-null object
national_waste_agency        159 non-null object
national_law_for_SW_mgmt     159 non-null object
PPP_rules_and_regulations    159 non-null object
population                   159 non-null float64
MSW_generated_tons_year      159 non-null float64
compost_percent              159 non-null float64
recycling_percent            159 non-null float64
gdp_inhab                    159 non-null float64
dt

In [54]:
countries_df_final = countries_df_4.drop(waste_compo_var+special_waste_var,axis=1)
countries_df_final.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 215 entries, 0 to 216
Data columns (total 12 columns):
country_code                 215 non-null object
region_id                    215 non-null object
income_id                    215 non-null object
info_system_SW_mgmt          215 non-null object
national_waste_agency        215 non-null object
national_law_for_SW_mgmt     215 non-null object
PPP_rules_and_regulations    215 non-null object
population                   215 non-null float64
MSW_generated_tons_year      215 non-null float64
compost_percent              215 non-null float64
recycling_percent            215 non-null float64
gdp_inhab                    215 non-null float64
dtypes: float64(5), object(7)
memory usage: 21.8+ KB


#### So finally we have divided rough countries dataset into the following datasets:

 - ***countries_df_final***: this dataset contains countries information, waste generation and existence of recycling for 215 countries. We will use it to predict waste generation in function of income and other dataset parameters.
 
 - ***countries_waste_composition_df***: it will allow to check the composition of waste considering the following 6 components: organic, glass, metal, paper/cardnoard, plastic and others. We have the information for 159 countries.
 
 - ***countries_construction_waste_df***, ***countries_e_waste_df***, ***countries_hazardous_waste_df*** and ***countries_medical_waste*** are the datsaets containing information about special waste. They contain information about 74, 182, 118 and 105 countries respectively.
 

#### Finally we export to csv files all these datasets:

In [55]:
countries_df_final.to_csv('CountriesWasteGeneration.csv', index=False)

countries_waste_composition_df.to_csv('CountriesWasteComposition.csv', index=False)

countries_construction_waste_df.to_csv('CountriesConstructionWaste.csv',index=False)
countries_e_waste_df.to_csv('CountriesEWaste.csv',index=False)
countries_hazardous_waste_df.to_csv('CountriesHazardousWaste.csv',index=False)
countries_medical_waste.to_csv('CountriesMedicalWaste.csv',index=False)